### ISSUE TITLE
"signed 32-bit int cannot represent all possible unsigned 2-bit values"?

### ISSUE THREAD

**Issue by donaldcallen** [ziglang/zig#18411]
### Zig Version

0.12.0-dev.1859+27d4bf753

### Steps to Reproduce and Observed Behavior

Build this
```
const std = @import("std");
const warn = std.debug.print;

fn baz(n:*const i32) void {
    warn("{d}\n", .{n.*});
}

pub fn main() void {
    const foo = enum {
        a,
        b,
        c
    };

    const b =  @intFromEnum(foo.b);
    baz(&b);
}
```
results in
```
dca@giovanni:/home/dca/Software/zig_tests$ zig build-exe test8.zig
test8.zig:16:9: error: expected type '*const i32', found '*const u2'
    baz(&b);
        ^~
test8.zig:16:9: note: pointer type child 'u2' cannot cast into pointer type child 'i32'
test8.zig:16:9: note: signed 32-bit int cannot represent all possible unsigned 2-bit values
test8.zig:4:10: note: parameter type declared here
fn baz(n:*const i32) void {
         ^~~~~~~~~~
referenced by:
    callMain: /home/dca/bin/Linux/lib/std/start.zig:575:17
    initEventLoopAndCallMain: /home/dca/bin/Linux/lib/std/start.zig:519:34
    remaining reference traces hidden; use '-freference-trace' to see all reference traces
```
Change the program thusly:
```
const std = @import("std");
const warn = std.debug.print;

fn baz(n:*const i32) void {
    warn("{d}\n", .{n.*});
}

pub fn main() void {
    const foo = enum {
        a,
        b,
        c
    };

    baz(&@intFromEnum(foo.b));
}
and it compiles without error. Running it:
```
```
dca@giovanni:/home/dca/Software/zig_tests$ zig build-exe test9.zig
dca@giovanni:/home/dca/Software/zig_tests$ ./test9
1
```
Aside from the fact that I would expect the first version to behave as the second does,
the error message "signed 32-bit int cannot represent all possible unsigned 2-bit values" is just wrong.

### Expected Behavior

I would expect the first version to behave as the second does.

**Comment by Vexu** [comment]
Duplicate of #13586

**Comment by donaldcallen** [comment]
I think you closed this without considering that there are two bugs I've run across:

1. The two version of the example I provided should produce identical results. They do not.
2. The error message is nonsense.

I think you focused on 2.

I went to some trouble to provide you with this report (I ran across the issue in real code and condensed it into examples for you, which took some time and effort) and it is more than a bit discouraging that my report was closed unjustifiably, in my opinion. The likelihood of my doing this again has gone down considerably as a result.

**Comment by Vexu** [comment]
In the first example you take the address of a `u2` variable while in the second the result location has type `*const i32` so the result of `@intFromEnum` is casted to the proper type.

**Comment by donaldcallen** [comment]
I do not understand your explanation. In the first example, the type of b is u2 and foo is called with &b, the address of a u2. 

In the second example, &@intFromEnum(foo.b) is the argument to foo. The documentation for @intFromEnum says "Converts an enumeration value into its integer tag type", which is u2. So if the documentation can be believed, in both cases the address of a u2 is being passed to the *const i32 parameter. So the cases look the same to someone who is not familiar with the gory details of the compiler’s internals. In other words, a user.

But you say the cases are different. Can you point me to where this is documented?

I might also mention that clang and gcc treat these two cases identically in C. Neither is allowed, which is fine. At least what seems like it ought to be the same is the same.

**Comment by Vexu** [comment]
> But you say the cases are different. Can you point me to where this is documented?

It's not documented yet but #18043 might explain it.

With a debug build you can see the difference using `zig ast-check -t`:
```
        %42 = field_val(%38, "b")
        %43 = int_from_enum(%42)
        %50 = ref(%43)
        %48 = call(nodiscard .auto, %46, [
          {
            %49 = validate_ref_ty(%48)
            %51 = break_inline(%48, %50)
          },

vs

        %43 = call(nodiscard .auto, %41, [
          {
            %46 = field_val(%38, "b")
            %47 = int_from_enum(%46)
            %48 = coerce_ptr_elem_ty(%43, %47)
            %49 = ref(%48)
            %50 = break_inline(%43, %49)
          },
```

> I might also mention that clang and gcc treat these two cases identically in C. Neither is allowed, which is fine. At least what seems like it ought to be the same is the same.

Zig isn't C.

**Comment by donaldcallen** [comment]
"Zig isn't C". I don't need to be told that different languages are different. 

You have completely missed my point in raising how this is handled in C, so I will re-state what I said somewhat differently, hoping for a better outcome: the two cases look identical to this user and I am quite confident that this will be true of others as well. My  purpose in raising the example of C is that there, these similar cases are handled consistently, as I would expect.

That they are different in Zig needs to be documented. It is far from sufficient to show me how the AST differs in the two cases unless your intention is that Zig will be used only by compiler writers. I asked the question about whether the difference in these two cases was documented because I was quite sure the answer would be "no", which is the answer I got. I have said this before and I will say it again -- this language is very poorly documented, 9 years after the inception of the project. Even though it has not reached v1.0 in all those years, the lack of good documentation is a detriment to this project, because you are driving away capable people willing to try beta-level software. Without good documentation, they can't know what it is *supposed* to do, so they can't tell you when it falls short, which is what you need. That's *exactly* what happened in this case.

I would also like to point out to you that you closed my issue without any explanation whatsoever of why you were dismissing what we are discussing here. You provided only a link to another issue that addresses the erroneous error message I got, stating that 32-bit i32s can't represent all values of a u2. It was only after I wrote back that you explained why you dismissed the first of the two problems I reported. And that explanation remains unconvincing.

I am giving you this feedback because I am the user you want -- an experienced software developer beta testing your work who is willing to provide useful feedback. Your responses, and the continuing lack of adequate documentation, are a major disincentive to continue doing that. And the whole tone of this interchange leads me to wish you good luck with the project; I won't be back.
